docs/guides/adding-component-types.md

```diff
@@ -0,0 +1,133 @@
+`@carbon/react`.
+
+<!-- prettier-ignore-start -->
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+## Table of Contents
+
+- [Goal](#goal)
+  - [Purpose](#purpose)
+- [Strategy](#strategy)
+  - [Steps to provide baseline type definitions for components](#steps-to-provide-baseline-type-definitions-for-components)
+- [FAQ](#faq)
+  - [How do I know what's part of the public api?](#how-do-i-know-whats-part-of-the-public-api)
+  - [Should components have both prototypes and ts interface?](#should-components-have-both-prototypes-and-ts-interface)
+  - [Should comment docs be duplicated into the ts interface?](#should-comment-docs-be-duplicated-into-the-ts-interface)
+  - [Where should I put the ts interface in the file?](#where-should-i-put-the-ts-interface-in-the-file)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+<!-- prettier-ignore-end -->
+
+## Goal
+
+The goal of this workstream is to provide as much downstream value to consumers
+who are using TypeScript, as quickly as possible, while writing the least amount
+of TypeScript.
+
+The goal is _not_ to convert the entire codebase to TypeScript right now.
+
+### Purpose
+
+By adding TypeScript types to components we anticipate a number of benefits:
+
+- Developer productivity will increase due to Component API's being
+  self-documenting and providing tight integration with code editor
+  intellisense.
+- The qualtiy of products developed will increase due to more stable, correct,
+  and thorough component API typings provided first-party through
+  `@carbon/react` itself.
+- Maintenance of the types themselves will be simplified by not having to go
+  through the DefinetlyTyped contribution process/system.
+
+Despite these benefits, adding TypeScript to the codebase is still a large shift
+for the developer community surrounding the Carbon Design System. Most projects
+are not using TypeScript, and contributors are more likely to not have in-depth
+knowledge of TypeScript.
+
+The bar to contribute to Carbon should be as low as possible to facilitate
+experimentation, innovation, and progress within the system. Adding TypeScript
+raises this bar of contribution adding additional friction to even the smallest
+of pull requests.
+
+We believe the majority of the benefits of TypeScript can be provided to
+consumers without needing to convert the entire codebase to use TypeScript.
+Overall, this effort will be focused on limiting the amount of TypeScript within
+the repository for now.
+
+## Strategy
+
+TypeScript will be incrementally adopted, focusing first on adding types to
+components prop APIs that are included as part of the public API of
+`@carbon/react`.
+
+Internal components, helpers, function, etc. will not initially be typed. These
+internal files/components not included in the public API should be left as `.js`
```

Comment:
Would it be useful to utilize [`// @ts-check`](https://www.typescriptlang.org/docs/handbook/intro-to-js-ts.html#ts-check) for files left with the `.js` extension?

Reply:
Yeah good call, I'll add it in. Links to that documentation would be good as well.